Openmarket sms integration

ricardomarquesramos opened the pull request:
This PR will add functionality to the sms_broker gem, to allow it to integrate with the OpenMarket SMS Provider.

We'll support two basic endpoints, the send SMS and check the delivery status of a message.

Inside this PR is also, some code that could be extracted as a standalone API wrapper for OpenMarket, but for now it comes bundled inside the sms_broker gem.

x1s commented on lib/open_market/sms/responses/message_status.rb:
```diff
@@ -0,0 +1,82 @@
+require 'pry'
```
will it break in production?

x1s commented on lib/open_market/sms/responses/send_message.rb:
```diff
@@ -0,0 +1,78 @@
+require 'pry'
```
same here

x1s commented on lib/sms_broker/client/response/error.rb:
```diff
@@ -27,3 +27,4 @@ def invalid_sender_id?
 require 'sms_broker/client/response/nexmo_error'
 require 'sms_broker/client/response/twilio_error'
 require 'sms_broker/client/response/nexmo_voice_error'
+require 'sms_broker/client/response/open_market/error'
```
can we have a empty line at the end?

x1s commented on lib/sms_broker/client/response/message_status_success.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+end
+
+require 'sms_broker/client/response/open_market/message_status_success'
```
any reason for the required be at the end?

ricardomarquesramos replied:
Just following the pattern the gem already had

jfpimentel commented on lib/open_market/sms/requests/send_message.rb:
```diff
@@ -0,0 +1,54 @@
+        end
+
+        def source
+          { source: { ton: 5, address: @from } }
```
5 could be extracted to a constant. Does it mean alphanumeric type of number?

jfpimentel commented on lib/open_market/sms/responses/message_status.rb:
```diff
@@ -0,0 +1,82 @@
+          "Response: #{status}, #{error}"
+        end
+
+        def error
```
Would change to `def error_message` to avoid confusing it with the error hash and to be consistent with `error_code`. Same case in `SendMessage` class.

ricardomarquesramos replied:
👍

jfpimentel commented on lib/open_market/sms/responses/send_message.rb:
```diff
@@ -0,0 +1,78 @@
+
+        def success?
+          # This can be improved.
+          status < 300
```
Notice the comment: should this be improved?

ricardomarquesramos replied:
👍

jfpimentel commented on lib/open_market/sms/responses/send_message.rb:
```diff
@@ -0,0 +1,78 @@
+          return unless success?
+
+          location = @response.headers['location']
+          *, @message_id = location.split('/')
```
I didn't know we could do this in ruby, so thanks 👍

ricardomarquesramos replied:
neat trick =)

ricardomarquesramos commented:
@jfpimentel addressed your points on the review.

Also @x1s somehow git is not assuming the extra line I've put at the end of that file =( but I did add it

jfpimentel commented on lib/open_market/sms/requests/send_message.rb:
```diff
@@ -44,7 +46,7 @@ def destination
 
         def source
-          { source: { ton: 5, address: @from } }
+          { source: { ton: TON_CODE, address: @from } }
```
I'd go one step further and be more explicit by calling it `ALPHANUMERIC_TON_CODE` or even `ALPHANUMERIC_TYPE_OF_NUMBER`. There isn't much of a difference between `ton: 5` and `ton: TON_CODE`. When I first saw it I was like "What's ton?" and I had to google "ton" and check the open market docs to understand what "ton" and 5 meant. But feel free to move on with it the way it is if you think I'm being too picky. It's fine 😉

ricardomarquesramos replied:
Makes perfect sense
